homeassistant/components/crownstone/entry_manager.py

```diff
@@ -0,0 +1,226 @@
+                    persistent_notification.async_create(
+                        self.hass,
+                        f"Setup of Crownstone USB dongle was unsuccessful on port {port}.\n \
+                        Crownstone Cloud will be used to switch Crownstones.\n \
+                        Please check if your port is correct and set up the USB again from integration options.",
+                        "Crownstone",
+                        "crownstone_usb_dongle_setup",
+                    )
+                # Create listeners for uart events
+                setup_uart_listeners(self)
+
+        # Save in what sphere the Crownstone USB is if it was setup correctly
+        # Crownstones could only be switched using the USB if they are in the same sphere (using BLE)
+        # Assuming that the other spheres are different buildings
+        if hasattr(self, ATTR_UART):
+            for sphere in self.cloud.cloud_data:
+                for crownstone in sphere.crownstones:
+                    if crownstone.type == CROWNSTONE_USB:
+                        self.usb_sphere_id = sphere.cloud_id
+
+        # create listener for when home assistant is stopped
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.on_shutdown)
+
+        # register all entities
+        self.hass.config_entries.async_setup_platforms(self.config_entry, PLATFORMS)
+
+        return True
+
+    async def async_process_events(self, sse_client: CrownstoneSSEAsync) -> None:
+        """Asynchronous iteration of Crownstone SSE events."""
+        async with sse_client as client:
+            async for event in client:
+                if event is not None:
+                    # make SSE updates, like ability change, available to the user
+                    self.hass.bus.async_fire(f"{DOMAIN}_{event.type}", event.data)
+
+    async def async_unload(self) -> bool:
+        """Unload the current config entry."""
+        # authentication failed
+        if self.cloud.cloud_data is None:
+            return True
+
+        # close sse client and unsub from listeners
+        self.sse.close_client()
+        for sse_unsub in self.listeners[SSE_LISTENERS]:
+            sse_unsub()
+
+        # close uart connection and unsub from listeners
+        if hasattr(self, ATTR_UART):
+            self.uart.stop()
+            for subscription_id in self.listeners[UART_LISTENERS]:
+                UartEventBus.unsubscribe(subscription_id)
+
+        # unload all platform entities
+        unload_ok = await self.hass.config_entries.async_unload_platforms(
+            self.config_entry, PLATFORMS
+        )
+
+        if unload_ok:
+            self.hass.data[DOMAIN].pop(self.config_entry.entry_id)
+
+        return unload_ok
+
+    @callback
+    def on_shutdown(self, _: Event) -> None:
+        """Close all IO connections."""
+        self.sse.close_client()
+        if hasattr(self, ATTR_UART):
+            self.uart.stop()
+
```

Comment:
Is this safe to run in the event loop. Does it do I/O?

Reply:
This simply calls `stop` for all the running threads which ends them https://github.com/crownstone/crownstone-lib-python-uart/blob/master/crownstone_uart/core/CrownstoneUart.py#L125

Not sure if this is blocking but i could also wrap it in an excecutor just to be safe, if you agree to it.

Reply:
If it's calling .join under the hood, that can block and needs to run in the excecutor

Reply:
I've traced it and `.join()` is called anywhere, it's simply setting the flags for the running functions to break out of their loop